drivers/xen/regions.c

```diff
@@ -0,0 +1,191 @@
+ */
+static struct extended_regions ext_blocks[DT_NUM_REGS(XEN_HYP_NODE)] = {
+	DT_FOREACH_REG(XEN_HYP_NODE, EXT_REG_POINTER)
+};
+
+void *xen_region_get_pages(size_t nr_pages)
+{
+	int ret, i;
+	void *ptr;
+
+	/*
+	 * Locking ranges iteration, although each call sys_mem_blocks_alloc_contiguous
+	 * has it's own syncronisation, to avoid the situation of 2 parralel calls of
+	 * get_page and put_page on full buffers when put_page clears space in block 1
+	 * when get_page has already checked block 1. This will lead us to -ENOMEM while
+	 * block 1 space was freed already
+	 */
+	k_mutex_lock(&ext_regions_lock, K_FOREVER);
+
+	for (i = EXTENDED_REGIONS_IDX; i < DT_NUM_REGS(XEN_HYP_NODE); i++) {
+		ret = sys_mem_blocks_alloc_contiguous(ext_blocks[i].block, nr_pages, &ptr);
+		if (ret == 0) {
+			break;
+		} else if (ret && ret != -ENOMEM) {
+			LOG_ERR("Unable to allocate pages, ret = %d\n", ret);
+			ptr = NULL;
+		}
+	}
+
+	k_mutex_unlock(&ext_regions_lock);
+
+	return ptr;
+}
+
+static int addr_is_in_buffer(uint8_t *ptr, sys_mem_blocks_t *block)
+{
+	/* Compare ptr to buffer address to determine if it was allocated in this block */
+	return (ptr >= block->buffer) &&
+		(ptr < block->buffer + (block->num_blocks << block->blk_sz_shift));
+}
+
+void xen_region_put_pages(void *ptr, size_t nr_pages)
+{
+	int ret, i;
+	sys_mem_blocks_t *block = NULL;
+
+	k_mutex_lock(&ext_regions_lock, K_FOREVER);
+
+	for (i = EXTENDED_REGIONS_IDX; i < DT_NUM_REGS(XEN_HYP_NODE); i++) {
+		if (addr_is_in_buffer(ptr, ext_blocks[i].block)) {
+			block = ext_blocks[i].block;
+		}
+	}
+
+	if (!block) {
+		LOG_ERR("Unable to find buffer, matching to addr %p\n", ptr);
+		goto unlock;
+	}
+
+	ret = sys_mem_blocks_free_contiguous(block, ptr, nr_pages);
+	if (ret) {
+		LOG_ERR("Unable to free pages, ret = %d\n", ret);
+		goto unlock;
+	}
+
+unlock:
+	k_mutex_unlock(&ext_regions_lock);
+}
+
+static int xen_regions_init(const struct device *d)
+{
+	int i;
+
+	ARG_UNUSED(d);
+
+	LOG_DBG("Initializing Xen Extended regions driver");
+
+	k_mutex_lock(&ext_regions_lock, K_FOREVER);
+
+	for (i = EXTENDED_REGIONS_IDX; i < DT_NUM_REGS(XEN_HYP_NODE); i++) {
+		/*
+		 * Using arch_mem_map to map the extended regions as 1:1.
+		 * The main idea of this functionality is to avoid using heap
+		 * and use 'unused memory' provided by XEN as extended_regions.
+		 * If using device_map call to map the space - then there is no
+		 * difference comparing to heap - we will get vaddr.
+		 * The idea is to allocate memory for the foreign mapping outside
+		 * of the region Z_VIRT_RAM_START..Z_VIRT_RAM_END.
+		 * phys_addr and size variables must be aligned to
+		 * CONFIG_MMU_PAGE_SIZE.
+		 */
+		arch_mem_map((void *)ext_blocks[i].phys_addr, ext_blocks[i].phys_addr,
+			     ext_blocks[i].size, K_MEM_CACHE_WB | K_MEM_PERM_RW);
+	}
+
+	k_mutex_unlock(&ext_regions_lock);
+
+	return 0;
+}
+
+SYS_INIT(xen_regions_init, POST_KERNEL, CONFIG_KERNEL_INIT_PRIORITY_DEFAULT);
```

Comment:
Some drivers, that can use extended regions may be initialized much earlier, than `POST_KERNEL`. May be initialize it on some `PRE_KERNEL_*` stage? I don't see why it is not possible